org-eclipse-che-jdt-ls-extension/jdt.ls.extension.core/src/main/java/org/eclipse/che/jdt/ls/extension/core/internal/externallibrary/LibraryNavigation.java

```diff
@@ -0,0 +1,655 @@
+    List<JarEntry> result = new ArrayList<>();
+    for (Object root : rootContent) {
+      if (root instanceof IPackageFragment) {
+        IPackageFragment packageFragment = (IPackageFragment) root;
+        JarEntry entry =
+            new JarEntry(
+                getSpecificText((IJavaElement) root),
+                packageFragment.getElementName(),
+                PACKAGE_ENTRY_TYPE);
+        result.add(entry);
+      }
+
+      if (root instanceof IClassFile) {
+        IClassFile classFile = (IClassFile) root;
+        JarEntry entry =
+            new JarEntry(
+                classFile.getElementName(),
+                classFile.findPrimaryType().getFullyQualifiedName(),
+                CLASS_FILE_ENTRY_TYPE);
+        result.add(entry);
+      }
+
+      if (root instanceof JarEntryResource) {
+        result.add(getJarEntryResource((JarEntryResource) root));
+      }
+    }
+    result.sort(comparator);
+    return result;
+  }
+
+  private static JarEntry getJarEntryResource(JarEntryResource resource) {
+    JarEntry entry = new JarEntry();
+    if (resource instanceof JarEntryDirectory) {
+      entry.setEntryType(FOLDER_ENTRY_TYPE);
+    }
+    if (resource instanceof JarEntryFile) {
+      entry.setEntryType(FILE_ENTRY_TYPE);
+    }
+    entry.setName(resource.getName());
+    entry.setPath(resource.getFullPath().toOSString());
+    return entry;
+  }
+
+  private static String getSpecificText(IJavaElement element) {
+    if (element instanceof IPackageFragment) {
+      IPackageFragment fragment = (IPackageFragment) element;
+      IJavaElement parent = getHierarchicalPackageParent(fragment);
+      if (parent instanceof IPackageFragment) {
+        return getNameDelta((IPackageFragment) parent, fragment);
+      }
+    }
+
+    return JavaElementLabels.getElementLabel(element, 0);
+  }
+
+  private static String getNameDelta(IPackageFragment parent, IPackageFragment fragment) {
+    String prefix = parent.getElementName() + '.';
+    String fullName = fragment.getElementName();
+    if (fullName.startsWith(prefix)) {
+      return fullName.substring(prefix.length());
+    }
+    return fullName;
+  }
+
+  private static IJavaElement getHierarchicalPackageParent(IPackageFragment child) {
+    String name = child.getElementName();
+    IPackageFragmentRoot parent = (IPackageFragmentRoot) child.getParent();
+    int index = name.lastIndexOf(PERIOD);
+    if (index != -1) {
+      String realParentName = name.substring(0, index);
+      IPackageFragment element = parent.getPackageFragment(realParentName);
+      if (element.exists()) {
+        try {
+          if (isEmpty(element) && findSinglePackageChild(element, parent.getChildren()) != null) {
+            return getHierarchicalPackageParent(element);
+          }
+        } catch (JavaModelException e) {
+          // ignore
+        }
+        return element;
+      }
+    }
+    return parent;
+  }
+
+  private static Comparator<JarEntry> comparator =
```

Comment:
This is a constant, right? Why not make it upper case and move it to the top of the file?

Reply:
It's right. Fixed